# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/html/HTMLSourceElement.cpp
# Commit: acdd013202e1
# Full Hash: acdd013202e1e688986e7e6c747b7c7fee38474e
# Author: Boris Chiou <boris.chiou@gmail.com>
# Date: 2022-08-23 02:46:57
# Regressor Bug: 1694741
# File Overlap Count: 1
# Description:
#   Bug 1694741 - Part 6: Map width/height attributes to the style of img elements. r=emilio
#   
#   This patch will use the width/height attributes from <source> to override
#   width/height/aspect-ratio CSS property values of <img> elements.
#   
# ==============================================================================

diff -r 2b4d9516b882 -r acdd013202e1 dom/html/HTMLSourceElement.cpp
--- a/dom/html/HTMLSourceElement.cpp	Mon Aug 22 20:18:39 2022 +0000
+++ b/dom/html/HTMLSourceElement.cpp	Mon Aug 22 20:18:39 2022 +0000
@@ -14,11 +14,13 @@
 #include "mozilla/dom/MediaList.h"
 #include "mozilla/dom/MediaSource.h"
 
-#include "nsGkAtoms.h"
-
 #include "mozilla/dom/BlobURLProtocolHandler.h"
 #include "mozilla/Preferences.h"
 
+#include "nsGkAtoms.h"
+#include "nsHTMLStyleSheet.h"
+#include "nsMappedAttributes.h"
+
 NS_IMPL_NS_NEW_HTML_ELEMENT(Source)
 
 namespace mozilla::dom {
@@ -131,6 +133,17 @@
         NS_GetSourceForMediaSourceURI(uri, getter_AddRefs(mSrcMediaSource));
       }
     }
+  } else if (StaticPrefs::dom_picture_source_dimension_attributes_enabled() &&
+             aNameSpaceID == kNameSpaceID_None &&
+             IsAttributeMappedToImages(aName) && IsInPicture()) {
+    BuildMappedAttributesForImage();
+
+    nsCOMPtr<nsIContent> sibling = AsContent();
+    while ((sibling = sibling->GetNextSibling())) {
+      if (auto* img = HTMLImageElement::FromNode(sibling)) {
+        img->PictureSourceDimensionChanged(this, aNotify);
+      }
+    }
   }
 
   return nsGenericHTMLElement::AfterSetAttr(
@@ -146,12 +159,77 @@
     media->NotifyAddedSource();
   }
 
+  if (aParent.IsHTMLElement(nsGkAtoms::picture)) {
+    BuildMappedAttributesForImage();
+  } else {
+    mMappedAttributesForImage = nullptr;
+  }
+
   return NS_OK;
 }
 
+void HTMLSourceElement::UnbindFromTree(bool aNullParent) {
+  mMappedAttributesForImage = nullptr;
+  nsGenericHTMLElement::UnbindFromTree(aNullParent);
+}
+
 JSObject* HTMLSourceElement::WrapNode(JSContext* aCx,
                                       JS::Handle<JSObject*> aGivenProto) {
   return HTMLSourceElement_Binding::Wrap(aCx, this, aGivenProto);
 }
 
+void HTMLSourceElement::BuildMappedAttributesForImage() {
+  MOZ_ASSERT(NS_IsMainThread());
+
+  if (!StaticPrefs::dom_picture_source_dimension_attributes_enabled()) {
+    return;
+  }
+
+  mMappedAttributesForImage = nullptr;
+
+  Document* document = GetComposedDoc();
+  nsHTMLStyleSheet* sheet =
+      document ? document->GetAttributeStyleSheet() : nullptr;
+  if (!sheet) {
+    return;
+  }
+
+  const nsAttrValue* width = mAttrs.GetAttr(nsGkAtoms::width);
+  const nsAttrValue* height = mAttrs.GetAttr(nsGkAtoms::height);
+  if (!width && !height) {
+    return;
+  }
+
+  const size_t count = (width ? 1 : 0) + (height ? 1 : 0);
+  RefPtr<nsMappedAttributes> modifiableMapped(new (count) nsMappedAttributes(
+      sheet, nsGenericHTMLElement::MapPictureSourceSizeAttributesInto));
+  MOZ_ASSERT(modifiableMapped);
+
+  auto maybeSetAttr = [&](nsAtom* aName, const nsAttrValue* aValue) {
+    if (!aValue) {
+      return;
+    }
+    nsAttrValue val(*aValue);
+    bool oldValueSet = false;
+    modifiableMapped->SetAndSwapAttr(aName, val, &oldValueSet);
+  };
+  maybeSetAttr(nsGkAtoms::width, width);
+  maybeSetAttr(nsGkAtoms::height, height);
+
+  RefPtr<nsMappedAttributes> newAttrs =
+      sheet->UniqueMappedAttributes(modifiableMapped);
+  NS_ENSURE_TRUE_VOID(newAttrs);
+
+  if (newAttrs != modifiableMapped) {
+    // Reset the stylesheet of modifiableMapped so that it doesn't
+    // spend time trying to remove itself from the hash.  There is no
+    // risk that modifiableMapped is in the hash since we created
+    // it ourselves and it didn't come from the stylesheet (in which
+    // case it would not have been modifiable).
+    modifiableMapped->DropStyleSheetReference();
+  }
+
+  mMappedAttributesForImage = std::move(newAttrs);
+}
+
 }  // namespace mozilla::dom